cooperator/models/subscription_request.py

```diff
@@ -78,54 +80,60 @@ def _check_share_available_to_user(self):
     def _send_confirmation_mail(self):
         if self.company_id.send_confirmation_email:
             mail_template_notif = self.get_mail_template_notif(
-                is_company=self.partner_id.is_company
+                is_company=self.is_company
             )
             # sudo is needed to change state of invoice linked to a request
             #  sent through the api
             mail_template_notif.sudo().send_mail(self.id)
 
-    # todo code this in a more pythonic way
-    #  - do not bypass self.create in create_comp_sub_req
-    #  - use "not in" where applicable
-    #  - sanitize declaration and assignation of cooperator
-    #  beware if usage in investor_wallet_platform
-    @api.model
-    def create(self, vals):
-        partner_obj = self.env["res.partner"]
-        if not vals.get("partner_id"):
-            cooperator = False
-            if vals.get("email"):
-                cooperator = partner_obj.get_cooperator_from_email(vals.get("email"))
-            if cooperator:
-                vals["type"] = "subscription"
-                vals = self.is_member(vals, cooperator)
-                vals["partner_id"] = cooperator.id
+    def _find_partner_from_create_vals(self, vals):
+        """
+        Find the partner corresponding to the vals dict.
+
+        If partner_id is present in the dict, use it to find the partner.
+        Otherwise, search companies by register number and individuals by
+        email address. If a match is found, add partner_id in the dict.
+        """
+        partner_model = self.env["res.partner"]
+        partner_id = vals.get("partner_id")
+        if partner_id:
+            return partner_model.browse(partner_id)
+        if vals.get("is_company"):
+            partner = partner_model.get_cooperator_from_crn(
+                vals.get("company_register_number")
+            )
         else:
-            cooperator_id = vals.get("partner_id")
-            cooperator = partner_obj.browse(cooperator_id)
-            vals = self.is_member(vals, cooperator)
+            partner = partner_model.get_cooperator_from_email(vals.get("email"))
+        if partner:
+            vals["partner_id"] = partner.id
+        return partner
 
-        if not cooperator.cooperator:
-            cooperator.write({"cooperator": True})
+    @api.model
+    def create(self, vals):
+        partner = self._find_partner_from_create_vals(vals)
+        if partner:
+            pending_requests_domain = [
+                ("partner_id", "=", partner.id),
+                ("state", "in", ("draft", "waiting", "done")),
+            ]
+            # we don't use partner.coop_candidate because we want to also
+            # handle draft and waiting requests.
+            if partner.member or self.search(pending_requests_domain):
+                vals["type"] = "increase"
```

Comment:
I have the impression that self.search(pending_requests_domain) could return something even if the partner is not member.
And if the partner exists but is not member, the type is "new", not "increase" no ?

Reply:
the logic i used is that if a partner has already made a subscription request (even not validated yet, but not canceled), subsequent subscription requests should be of type `increase`. only the first subscription request would be of type `new`. does that sound logical to you?

Reply:
OK, It is a good option. It is not ideal because there can be cases where the first request will be cancelled later, and the only request will have type "increase". But I guess it is not very important...?

Reply:
Reading your comments, it looks like the type should ideally be attributed later, when the request is validated (not at creation), at which point it makes more sense to check whether the subscription is new of increase. Just saying, but probably ok to leave it as huguesdk proposes.

Reply:
very good remarks. the type on the request itself does not make much sense until it is validated. what is this type used for, anyway?